geopandas/io/file.py

```diff
@@ -428,22 +435,47 @@ def _read_file_pyogrio(path_or_bytes, bbox=None, mask=None, rows=None, **kwargs)
                 raise ValueError("slice with step is not supported")
         else:
             raise TypeError("'rows' must be an integer or a slice.")
+
+    if bbox is not None and mask is not None:
+        # match error message from Fiona
+        raise ValueError("mask and bbox can not be set together")
+
     if bbox is not None:
         if isinstance(bbox, (GeoDataFrame, GeoSeries)):
-            bbox = tuple(bbox.total_bounds)
+            crs = pyogrio.read_info(path_or_bytes).get("crs")
+            if isinstance(path_or_bytes, IOBase):
+                path_or_bytes.seek(0)
+
+            bbox = tuple(bbox.to_crs(crs).total_bounds)
         elif isinstance(bbox, BaseGeometry):
             bbox = bbox.bounds
         if len(bbox) != 4:
             raise ValueError("'bbox' should be a length-4 tuple.")
+
     if mask is not None:
-        raise ValueError(
-            "The 'mask' keyword is not supported with the 'pyogrio' engine. "
-            "You can use 'bbox' instead."
-        )
+        # NOTE: mask cannot be used at same time as bbox keyword
+        if not PYOGRIO_GE_07:
+            raise ValueError(
+                "The 'mask' keyword requires pyogrio >= 0.7.0.  "
+                "You can use 'bbox' instead."
+            )
+        if isinstance(mask, (GeoDataFrame, GeoSeries)):
+            crs = pyogrio.read_info(path_or_bytes).get("crs")
+            if isinstance(path_or_bytes, IOBase):
+                path_or_bytes.seek(0)
+
+            mask = shapely.unary_union(mask.to_crs(crs).geometry.values)
```

Comment:
Do we need the mask to be a result of union or can we just dump the whole GeoSeries to a single GeometryCollection?

Reply:
I'm not sure that we strictly need the union, and it is indeed a performance tradeoff.  In this case, I used `unary_union` to match the existing behavior for Fiona.  About the only case I can think of where the union may be functionally different than a GeometryCollection would be a set of overlapping polygons.  But provided the mask implementation within GDAL will return true of any of the overlapping polygons also intersect features in the dataset, it really shouldn't differ in behavior with respect to masking.

Perhaps we would run into issues if any of the geometries are already GeometryCollections though?  Since this would result in a GeometryCollection of GeometryCollections?

Reply:
Thanks for investigating that further @theroggy ; I did not consider the case of a GeometryCollection that contained overlapping polygons.  Sounds like we should stick to calculating the union.

Reply:
Shall we at least add a note about it to the docstring? The cost of union can be surprising and it could easily take way longer to get the mask we pass to GDAL than to do the actual IO.

Reply:
It is also possible to the other way around: mention that if an error occurs, the user also can do the union himself to avoid the error...

Reply:
@theroggy I am wondering if overlay operations actually matter in this specific case, as here it's for evaluating a predicate (it might still be that collections have less optimized implementations of the spatial predicates, but I wouldn't expect them to fail?)

Reply:
Not sure. I have never done specific tests on spatial predicates, so I don't know if their implementations are more tolerant. I did a quick test for the specific case above, and that seems to pass for "intersects" without error, so possibly those code paths indeed have already more general support for collections?

Reply:
Decision from last GeoPandas dev meeting was to add a note to the docstring to indicate that using mask will involve calculating the union, which can be slow, and then to do another PR at a later time to avoid the union in a consistent way for both Fiona and Pyogrio.